design/continuous-delivery.md

```diff
@@ -0,0 +1,329 @@
+# Continuous Delivery for CDK Apps
+
+This is a specification for continuous delivery support for CDK apps. It
+describes the requirements, proposed APIs and developer workflow.
+
+**_Goal_: full CI/CD for CDK apps at any complexity.**
+
+The desired developer experience is that teams will be able to “git push” a
+change into their CDK app repo and have this change automatically picked up,
+built, tested and deployed according to a deployment flow they define.
+
+Any changes in resources, assets *or stacks* in their apps will automatically be
+added. To that end, the deployment pipeline itself will also be continuously
+delivered and updated throughout the same workflow. No manual deployments to
+production (incl. the pipeline) will be required.
+
+The only caveat is that **new environments** (account/region) will need to be
+bootstrapped in advance in order to establish trust with the central pipeline
+environment and set-up resources needed for deployment such as the assets S3
+bucket.
+
+Prototype: https://github.com/eladb/cdkcd-test
+
+## Requirements
+
+This list describes only the minimal set of requirements from this feature.
+After we release these building blocks, we will look into vending higher-level
+"one liner" APIs that will make it very easy to get started.
+
+1. **Assets**: Support apps that include all supported kinds of assets (S3
+   files, ECR images)
+2. **Multiple-environments**: Support apps that have stacks that target multiple
+   environments (accounts/regions)
+3. **Orchestration**: Allow developers to express complex deployment
+   orchestration based on the capabilities of CodePipeline
+4. **User defined build + synth environment**: the runtime environment in which
+   the code is built and the CDK app is synthesized should be fully customizable
+   by the user
+5. **Controlled deployment environment**: for security reasons, the runtime
+   environment in which deployment is executed will be fully controlled and will
+   not allow running user code or user-defined image
+
+_Non-requirements/assumptions:_
+
+1. Bundling assets (creating Lambda zip files, building docker images) and
```

Comment:
I wonder if you've thought about having one CodeBuild step which publishes to all ECR repos (and/or relies on CP's cross region s3 copy) in the **build** stage of the Pipeline. This would mean folks could use native CloudFormation actions in CodePipeline (since the CDK step wouldn't be required to upload to ECR/S3 every step). 

I like that the native CloudFormation target support in CodePipeline gives you the ability to preview a stack change (create changeset then execute are two seperate actions). I love that the CDK is really good about this (showing me what I'm about to do), and I think for production stages, folks would be interested in maintaining the ability to do this. 

Additionally, with larger ECR images, it might be more cost efficient to have one CodeBuild action which builds and publishes.

Reply:
Will be addressed by `cdk-package`